### ISSUE TITLE
Switching between local/remote node is buggy sometimes.

### ISSUE THREAD

**Issue by xmrdsc** [monero-project/monero-gui#2206]
Switching between local/remote node or remote node to another remote node *can* be buggy. I say can because I don't know in what situation exactly it becomes buggy, but it is easy to reproduce by switching between remote nodes.

For example: 

1. go to Settings->Node
2. Enable Remote Node
3. Enter ip:port of node you find on `https://autonode.xmr.pm`
4. Press connect
5. It connects (hopefully)
6. Enter a different ip:port of a node
7. Press connect
8. It connects (hopefully)

Step 5 and 8 is where it can become problematic. Firstly, the user has no feedback of what the GUI is doing:

1. Is it busy connecting?
2. Was it able to do a TCP handshake (e.g: is the host reachable) ?
3. Is it a slow node? (perhaps it only started responding after 2000 msec
4. Did we connect?

Secondly, I believe libwalletqt (daemonManager) becomes quite confused if the remote node has certain kind of problems, at which point switching to another node and pressing 'Connect' will not do anything - as if the daemonManager is in limbo from the previous attempt(s). I could be wrong, then again there is no feedback to tell what is happening.

Would be nice if:

1. There is feedback, or a log to tell how the connection is progressing
2. A message when the connection failed
3. Verify that pressing 'Connect' will indeed destroy any previous attempt or state.
4. Maybe also introduce a 'Disconnect' button.

**Comment by Realchacal** [comment]
I can confirm I had the same issues a lot. As xmrdsc said, a main problem definitely is that you never know what the GUI currently is doing because the shown status/messages are not reliable enough. Example:

![image](https://user-images.githubusercontent.com/46682965/59216965-28236980-8bbd-11e9-8351-dea95399ff75.png)


(Hint: In the situation shown I have to press "Export View Key" on my Ledger Nano S - then everything is working fine. So I'm not looking for actual answers to my questions! It was meant as a hint for you, because such contradictory status/messages shouldn't be possible in the first place, I guess.)

**Comment by dEBRUYNE-1** [comment]
@Realchacal - Can you check again with GUI v0.14.1.0?

**Comment by Realchacal** [comment]
@dEBRUYNE-1 First let me say that it is great that the GUI now waits for the user to click on his hardware wallet "Export XPUP". This catches a lot of issues.

Node switching unfortunately is still buggy. What I just did:
1. Launch the new GUI version. It automatically connected to the remote node that I used the last time. Fine!
2. Enter another remote node address ("opennode.xmr-tw.org", same port)
3. Click connect. It disconnected! Nothing else.
4. Click connect again. Nothing.
5. Close wallet. It freezed for 10 seconds, then it actually closed.
6. Reopen the wallet. The remote node didn't connect automatically and pressing "connect" didn't help.
7. Switch to local node and instantly back to remote note. Didn't help.
7. Enter another remote node address ("node.moneroworld.com", same port), click connect. It connected! Fine!
9. Enter the remote node address from before again ("opennode.xmr-tw.org", same port), click connect. Now it works!

Apart from that: Status messages are still not good enough. You just don't know, what the GUI is doing. When switching between remote node and local node the status isn't updated immediately, instead I sometimes have to wait up to 2-3 seconds until I see a status change. This drives people crazy, especially if they have to wait much longer than 2-3 seconds because of bad internet connection. They just don't know "is it doing something? What is it doing? Do I have to press 'connect' or just wait?" There should be INSTANT status message changes when the user clicks something. Like "closing local node", "local node closed", "connecting to remote node (no response yet)", "connecting to remote node (is responding)" or something like that.

**Comment by MicahZoltu** [comment]
I am trying to connect to a remote node of mine and there is no information in either the GUI, the GUI logs tab, or the remote node's logs.  Clicking the connect button doesn't appear to do anything at all, though maybe it is trying behind the scenes and just giving zero feedback?  I don't know if the host is unreachable, or if my remote node is on an incompatible version (it is running 0.13.x while GUI is 0.14.x), or if there is some other problem.

I'm at a total loss as to why I can't connect, but at the moment I cannot use Monero and it is quite frustrating.  Some simple status/logs would go a _long_ way to making this situation better!

**Comment by selsta** [comment]
@MicahZoltu You can’t use v0.13 anymore.

Update both to v0.14.1.0 and it will work like expected.

**Comment by Realchacal** [comment]
> @MicahZoltu You can’t use v0.13 anymore.
> 
> Update both to v0.14.1.0 and it will work like expected.

Nope. I was using v0.14.1.0. (I thought that was obvious, since dEBRUYNE-1 asked me to check with v0.14.1.0 and I replied to that.)

**Comment by MicahZoltu** [comment]
I updated both to 0.14.1.0 and was _eventually_ able to connect.  Even after updating, I ran into many of the symptoms people above described, where the UI would appear to be stuck, closing would lock up the app for 10 seconds, etc.

**Comment by selsta** [comment]
@Realchacal I didn’t reply to you?

**Comment by Realchacal** [comment]
> @Realchacal I didn’t reply to you?

You said "update BOTH" that's why I thought you meant me, too :) Misunderstanding :)

**Comment by Realchacal** [comment]
I'd like to add:

- I'm using the default wallet mode. Switching from local node to remote node sometimes leads to "Waiting for daemon to stop..." and 20 seconds later (without doing anything) "Waiting for daemon to start..." After that, it says "Network status: Remote node."

**Comment by selsta** [comment]
> sometimes triggers no daemon stopping at all, it just says network status is "Remote node", while the daemon is still running.

This is the correct behaviour.

> sometimes leads to "Waiting for daemon to stop..." but after the daemon has stopped it then says "Waiting for daemon to start..." - although I haven't clicked anything.

Sounds like a bug, I’ve never experienced this. Are you pressing “Stop local node” before switching from local to remote node? Any way to reproduce?

**Comment by Realchacal** [comment]
> Sounds like a bug, I’ve never experienced this. Are you pressing “Stop local node” before switching from local to remote node? Any way to reproduce?

No, but I guess it may have something to do with changing modes too quickly, which may cause the node connections to be in some "half-established" state, similar to Schrödinger's cat.

**Comment by Realchacal** [comment]
- Be in Advanced mode, remote node
- Prepare a transaction on the `Send` page
- Switch to local node on the `Settings` page
- QUICKLY go back to the `Send` page
--> The `Send` button SHOULD be gray (because the daemon hasn't started yet) but it's orange and you can click it. Results in various unexpected behavior afterwards.

This is just one of many examples how the modes are messy, especially if you switch quickly between them. IMO we need #2320 as a first step, clearer network status messages #2304, and then some forced waiting times to prevent too quick mode switching.

**Comment by selsta** [comment]
This has been greatly improved in #2747
